Design doc: DeltaTree Read Thread Pool and Data Sharing (#5478)
pingcap/tiflash: docs/design/2022-07-25-read-thread-pool-and-data-sharing.md

# DeltaTree Read Thread Pool and Data Sharing

- Author: [Jinhe Lin](https://github.com/JinheLin)

## Table of Contents

* [Introduction](#introduction)
* [Motivation or Background](#motivation-or-background)
* [Detailed Design](#detailed-design)
* [Impacts & Risks](#impacts--risks)

## Introduction

DeltaTree is the columnar storage engine of TiFlash.

**Read Thread Pool** is a new component of DeltaTree that responsible for reading data and decoupling the computing layer and the storage layer.

At the same time, in order to improve the performance of multi-concurrent table scan, the threads of the storage layer will schedule and try to combine data read requests according to the characteristics of analytical requests, so that concurrent table scans on the same table can share data read operations and reduce duplication of work. We calls this feature **Data Sharing**.

## Motivation or Background

### Read Thread Model

> The read thread model of DeltaTree continues from ClickHouse, which is relatively simple and straightforward：

Reviewer: ```suggestion
The read thread model of DeltaTree inherits from ClickHouse, which is relatively simple and straightforward：
```

Data Sharing allows concurrent requests to share the reading data, reducing repetitive work such as I/O and decoding.

## Detailed Design

### Overview

From the high level of view, the reading of DeltaTree is divided into two steps:

1. Call `DeltaMergeStore::read` to create multiple `BlockInputStreams`s: obtain the `Segment`s to be read and create snapshots.

2. Call `BlockInputStream::read` to read data, one `Block` object at a time.

The detailed process is shown in the following figure:

![](./images/2022-07-25-read-thread-pool-and-data-sharing.png)

1. When calling `DeltaMergeStore::read` to create multiple `BlockInputStream`s, register the segments to be read to `SegmentReadTaskScheduler`.
2. The logic of the `BlockInputStream` is relatively simple, just pop the `Block` objects from a `Block` queue.
3. Scheduler thread is responsible for scheduling read tasks on `SegmentReadTaskScheduler`. During scheduling, concurrent read of the same segment will be merged, and read requests will be sent to the read thread pool through the queue for processing.
4. After the read thread pool reads the block from the segment, push it to the block queue.

### ReadThreadPool

> In order to avoid segments are processed accross NUMA nodes that affect performance :

Reviewer 2: ```suggestion
In order to avoid segments are processed across different NUMA nodes that affects performance :
```

1. Group the read threads according to NUMA nodes and bind them to the corresponding CPU.

> 2. The read task is hashed according to the segment's id, so it can be always processed by one NUMA nodes.

Reviewer 2: ```suggestion
2. The read task is hashed according to the segment's id, so it can be always processed by the same NUMA nodes.
```

### SegmentReadTaskScheduler

> 1. The purpose of `SegmentReadScheduler` is to allow the same segment of different read requests to be executed in the same time as much as possible to increase the opportunity for sharing the same data.

Reviewer 2: ```suggestion
1. The purpose of `SegmentReadScheduler` is to allow the same segment of different read requests to be executed at the same time as much as possible to increase the opportunity of sharing the same data.
```
